docs/proposals/petset.md

```diff
@@ -0,0 +1,414 @@
+  focus on ensuring the information necessary for the software to function is available. For example,
+  rather than providing a "post-creation" hook invoked when the cluster is complete, provide the necessary
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is preserved. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+A PetSet has 0..N **members**, each with a unique **identity** which is a name that is unique within the
+set.
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+		// A label selector that "owns" objects created under this set
+		Selector map[string]string
+
+    // If set, identities will be auto generated based on the name of the
+    // Set "NAME-$(identity)", i.e. "mysql-1", "mysql-2", ...
+    Members *int
+    // A structured set of identities that if set will be used to generate more
+    /// complex shard patterns
+    MemberIdentities []MemberIdentity
+}
+
+type MemberIdentity struct {
+    // a set of fixed identities, like ["mysql-master", "mysql-slave-1"]
+    Fixed []string
+    // a pattern for a range of identities
+    Range *IdentityRange
+    // a set of nested identities that are constructed from the parent pattern,
+    // allows nesting like "mysql-zone1-1", "mysql-zone1-2", "mysql-zone2-1"
+    Identities []MemberIdentity
+
+    // PROPOSED: Add a set of additional parameters per identity that can be
+		// injected into the templates
+}
```

Comment:
Zone sharding could just as easily be "role sharding" or "primary /
failover sharding", I chose zone to represent any sort of numbered shard
weight where the template is very similar.  Role isn't special per se, it's
probably an 80% case.  Rather than formalizing it, we should either have a
parameter or a name prefix that represents it.

Ultimately it comes down to copying the PetSet multiple times - I think
that as a general rule of thumb for small numbers of simple free parameters
(say <5) if you have to create multiple pet sets to satisfy the use case
then PetSet isn't very useful.  As an example, if you had ROLE=master and
ROLE=replica you might have a coupled parameter MEMORY=10GB and
MEMORY=1GB.  That suggests to me that MemberIdentity should have a (small)
set of key value param pairs.

If Role isn't special, and names *should* be meaningful to end users, we
still have to have some concept of either the pattern or range.  Min-Max
1=1 as a range isn't the end of the world, but it might also not be clear
to a user that it's "special".


On Tue, Mar 29, 2016 at 12:15 AM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/petset.md
> <https://github.com/kubernetes/kubernetes/pull/18016#discussion_r57670246>
> :
>
> > +    /// complex shard patterns
> > +    MemberIdentities []MemberIdentity
> > +}
> > +
> > +type MemberIdentity struct {
> > +    // a set of fixed identities, like ["mysql-master", "mysql-slave-1"]
> > +    Fixed []string
> > +    // a pattern for a range of identities
> > +    Range *IdentityRange
> > +    // a set of nested identities that are constructed from the parent pattern,
> > +    // allows nesting like "mysql-zone1-1", "mysql-zone1-2", "mysql-zone2-1"
> > +    Identities []MemberIdentity
> > +
> > +    // PROPOSED: Add a set of additional parameters per identity that can be
> > +		// injected into the templates
> > +}
>
>
>    1. Users able to control the names of the pets - names are important
>    2. Express 1 + 0..N ranges (master + read replicas)
>    3. Express zone sharding - cassandra-east-(1..10),
>    cassandra-west-(1..3)
>    4. Express simple "role" parameters per set for use in the template -
>    master gets ROLE=master and read replicas get ROLE=replica
>
> 3 feels like it belongs in a higher level resource. X:petset as
> deployment:RC and Ingress:service. Simply saying:
>
> type PetSetSpec struct {
>      Selector map[string]string
>      MemberIdentities []MemberIdentity
>      Template PodTemplateSpec
>  }
> type MemberIdentity struct {
>      Role string
>      mix, max int
>  }
>
> Would give us "cassandra cluster". Maybe we should start with that?
> Do we need 2 if we have 4?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18016/files/961f1f94c35d4979ac83bbad482090cb6c22781c#r57670246>
>